File: p134.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
        int n = gas.size();
        int curr = 0;
        for(int start = 0; start < n; ++start) {
            int curr_gas = 0;
            bool flag = true;
            curr = 0;
            while(curr < n + 1) {
                curr_gas += gas[(start+curr)%n];
                curr_gas -= cost[(start+curr)%n];
                 if(curr_gas < 0) {
                    flag = false;
                    break;
                }
                ++curr;
            }
            if(flag)
                return start;
        }
        return -1;
    }
};

File: p134.py
class Solution:
    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:
        n = len(gas)
        left, right = 0, 0

        curr = 0

        while True:
            flag = False
            while curr+gas[right%n]-cost[right%n] >= 0:
                curr += gas[right%n]-cost[right%n]
                right += 1
                if right-left > n:
                    flag = True
                    break
            while curr+gas[right%n]-cost[right%n] < 0:
                left -= 1
                curr += gas[left]-cost[left]
                if right-left >= n:
                    flag = True
                    break
            if flag:
                break

        return left%n if curr >= 0 else -1
